### PR TITLE
DX: Fix `defined` and `Check` types for CesiumJS devs

### PR DESCRIPTION
**Pull request by jjspace**
# Description

- Add `defined.d.ts` which will now get picked up and provide the correct type predicates and assertions so JS Type checking will understand things are defined after this call
    - Originally added in https://github.com/CesiumGS/cesium/pull/11455 which is fantastic for _external_ contributors but local development doesn't look at the generated `index.d.ts` file for types like it will with the new `defined.d.ts` file

| Before                                                                                               | After                                                                                                  |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
|  |  |

- Fixed `Check.d.ts` for local development so the TS server understands it's a default export not an exported object
    - This was originally added in https://github.com/CesiumGS/cesium/pull/11901 to help _external_ users of CesiumJS, primarily the ion team

| Before                                                                                               | After                                                                                                  |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
|  |  |

- Updated the build process to account for these files and changes. _The built `Cesium.d.ts` and `engine/index.d.ts` files should be identical to the previous ones_.

## Issue number and link

No issue, just a small DX issue I've grown too tired of seeing

## Testing plan

- In `main` run `npm run built-ts`
- Create a copy of `Source/Cesium.d.ts` and `packages/engine/index.d.ts` somewhere you can check later
- Switch to this branch and run `npm run build-ts`
- Do a diff of those files and ensure they are identical. This confirms our external use will be the same downstream
- To test locally find a usage of `defined` and `Check` locally
- Turn on the VSCode setting `Check JS` for type checking with the TS server (This may or may not be necessary, I'm not sure)
- Verify that in `main` the types aren't correct as shown above, verify that in this branch they are
- Turn off the `Check JS` setting if you're not ready to see a lot of other (false positive) errors scattered in our code

# Author checklist

- [ ] I have submitted a Contributor License Agreement
- [ ] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] I have updated the inline documentation, and included code examples where relevant
- [ ] I have performed a self-review of my code

### CODE REVIEW

File: gulpfile.js
Comment (jjspace) [inline gulpfile.js:1419]: As per my [comment](https://github.com/CesiumGS/cesium/pull/11901#discussion_r1543545338) when I did the same conversion into a `.d.ts` file I would _love_ if we could skip this `replace` and pass the types file into `tsd-jsdoc` but we cannot. I think `defined` and `Check` are some of the only types we really need to override to take advantage of stuff like type predicates but if we add more we should definitely revisit this setup as it will quickly become too unwieldy.

**Comment by jjspace** [comment]
@javagl Would you mind taking a look at this? I think you've been looking into our type system and typescript adjacent stuff recently so you might be a good person to provide input.

**Review by javagl — COMMENTED** [review]
I followed the points from the testing plan. There are some smaller caveats.

After enabling the "Check JS" functionality in VS Code, it **did** complain about many things. But it did **not** generate the "...is possibly undefined" messages for me.
Maybe my VS Code is somehow messed up...? Maybe I'd have to delete something or use "a fresh copy" of something...? More generally: _Where_ exactly is VSCode supposed to pull that new information from, when the `d.ts` files are supposed to be unchanged? 

It **did** complain about the `typeof` in `Check`. And this is resolved by this change. 
There may be an issue with the resulting `export default` in this case that likely will cause trouble in a TypeScript environment (see inlined comment). But this might be an easy fix.

---

Beyond that: 

This is **NOT** in the scope of this PR, but I cannot look at the `gulpfile.js` without pointing out that this is one of my pet peeves 😬 The fact that it does have to contain these `readFileSync(....)` and ridiculously obscure regexes is bad enough. The fact that it has to contain them **twice** makes it even worse. But when looking at that duplication like in a [blink comparator](https://en.wikipedia.org/wiki/Blink_comparator), and then noticing the difference in that `// inlined comment`, it is _nearly_ funny again 🙂 : 

Cesium Gulpfile duplication

Usually, I'm not touching that file, out of the fear of "breaking something". But ... after this is merged, I'll probably go out on a limb, take a pass at that file, and then there _might_ be a function that is called `fixUpTheOutputToMatchWhatWeNeed()` - I'm reasonably sure that this would not break anything, at least...

File: gulpfile.js
Comment (jjspace) [reply gulpfile.js]: The checked in `Check.d.ts` file _should_ have a newline at the end afaict but I also have no issue with making the newline optional in the regex. I've pushed up that change if you would like to take another look.

Comment (javagl) [reply gulpfile.js]: When you scroll to the bottom of https://github.com/CesiumGS/cesium/blob/5586cf1ab83c586fd1938ce29c24b12940a948de/packages/engine/Source/Core/Check.d.ts , does it display a newline for you?

Cesium Defined Check

I thought that one could alternatively look for the `;` (which kind of is a "line separator" anyhow). 
But making the `\n` optional seems to be sufficient, so that seems to be resolved now.

**Comment by javagl** [comment]
>  I sometimes find I have to run the Restart TS Server or Reload Window commands to refresh typescript.

Yeah, I also had cases where this was necessary. But it wasnt't the root cause here. Although I did search for `"defined("` and did click through _many_ cases, I didn't click to all (13509) appearances. And apparently, I only looked at ones that didn't show the message because there was no corresponding information in the JSDoc: **Only** when that "thing" is defined to be a non-`undefined` type via JSDoc, the message appears. For example, in this snippet...

```javascript
/**
 * Does not complain about s potentially being undefined
 */
function exampleA(s) {
  if (defined(s)) {
    console.log(s.length);
  }
}
/**
 * Does not complain about s potentially being undefined
 * 
 * @param {string} s
 */
function exampleB(s) {
  if (defined(s)) {
    console.log(s.length);
  }
}

/**
 * DOES complain about s potentially being undefined!
 * 
 * @param {string | undefined} s
 */
function exampleC(s) {
  if (defined(s)) {
    console.log(s.length);
  }
}
```

it shows the message for the bottommost one:

Cesium Defined Check 2

And I confirmed that this does no longer appear on the branch of this PR.

**Comment by jjspace** [comment]
@javagl glad you were able to confirm the issue and this fixes it for you too. Do you have any other comments on this?

**Review by javagl — APPROVED** [review]
@jjspace I assume that this was about the pending review approval...? 
Tested it and it works as intended. I think that it can be merged.

**Pull request merged**
